pydrofoil/bitvector.py

```diff
@@ -110,6 +110,18 @@ def rshift(self, i):
         if i >= self.size():
             return self.make(r_uint(0))
         return self.make(self.val >> i)
+    
+    def arith_shiftr(self, i):
+        assert i >= 0
+        if i == 0:
+            return self
+        if self.read_bit(self._size - 1) == 0:
+            return self.rshift(i)
+        val_new = (-self.val)/(2**i)
+        
+        return (SmallBitVector(self._size, -val_new) if (-self.val)%(2**i) == 0 else SmallBitVector(self._size, -val_new-1)) if i < self._size\
+        else SmallBitVector(self._size, (r_uint(1) << self._size) -1)
```

Comment:
The last three lines somehow feel too complicated for me. How about something like this (untested):

```python
if i > self._size:
    return SmallBitVector(self._size, (r_uint(1) << self._size) - 1) # all bits set
val = self.val >> i
extra_bits = ((1 << i) - 1) << (self._size - i) # set the highest i bits
return SmallBitVector(self._size, val | extra_bits)
```

`extra_bits` contains `i` bits that are set (that's the expression `((1 << i) - 1)`), and it's the bits that start from position `size - i`. Makes sense?

Once this works, maybe we can do the equivalent thing for the `GenericBitVector` variant and see if that works?

You can also look at the implementation of the `sign_extend` methods, they do something similar.